README.md
[![](https://jitpack.io/v/rjsuzuki/mopub-android-sdk.svg)](https://jitpack.io/#rjsuzuki/mopub-android-sdk)


# MoPub Android SDK

Thanks for taking a look at MoPub! We take pride in having an easy-to-use, flexible monetization solution that works across multiple platforms.

Sign up for an account at [http://app.mopub.com/](http://app.mopub.com/).

## Need Help?

You can find integration documentation on our [wiki](https://github.com/mopub/mopub-android-sdk/wiki/Getting-Started) and additional help documentation on our [developer help site](https://developers.mopub.com/publishers/android/).

To file an issue with our team visit the [MoPub Forum](https://twittercommunity.com/c/advertiser-api/mopub) or email [support@mopub.com](mailto:support@mopub.com).

## New Pull Requests?

Thank you for submitting pull requests to the MoPub Android GitHub repository. Our team regularly monitors and investigates all submissions for inclusion in our official SDK releases. Please note that MoPub does not directly merge these pull requests at this time. Please reach out to your account team or [support@mopub.com](mailto:support@mopub.com) if you have further questions.

## Download

The MoPub SDK is available via:

1. **Maven Central Repository**
    
    [ ![Download](https://api.bintray.com/packages/mopub/mopub-android-sdk/mopub-android-sdk/images/download.svg)](https://search.maven.org/search?q=MoPub-sdk)  
    The MoPub SDK is available as an AAR via Maven Central; to use it, add the following to your `build.gradle`.
    
    ```
    repositories {
        mavenCentral() // includes the MoPub SDK
        google() // necessary for Android API
    }

    dependencies {
        implementation('com.mopub:mopub-sdk:5.17.0@aar') {
            transitive = true
        }
    }
    ```

    ***SDK Modularization***

    With the modular SDK, you can choose to include specific formats to decrease overall SDK footprint in your app. To do so, include the line for any combination of components that you want in your `build.gradle` file as follows:

    ```groovy
    repositories {
        // ... other project repositories
        mavenCentral() // includes the MoPub SDK
        google() // necessary for Android API
    }

    dependencies {
        // ... other project dependencies

        // For banners
        implementation('com.mopub:mopub-sdk-banner:5.17.0@aar') {
            transitive = true
        }
        
        // For interstitials and rewarded ads
        implementation('com.mopub:mopub-sdk-fullscreen:5.17.0@aar') {
            transitive = true
        }

        // For native static (images).
        implementation('com.mopub:mopub-sdk-native-static:5.17.0@aar') {
            transitive = true
        }
    }
    ```

    **To continue integration using the mopub-sdk AAR, please see the [Getting Started guide](https://github.com/mopub/mopub-android-sdk/wiki/Getting-Started#updating-your-android-manifest).**

2. **Zipped Source**

    The MoPub SDK is also distributed as zipped source code that you can include in your application:

    **[MoPub Android SDK.zip](http://bit.ly/YUdWhH)**  
    _Includes everything you need to serve MoPub ads.  No third party ad networks are included._
    
    **For additional integration instructions, please see the [Getting Started guide](https://github.com/mopub/mopub-android-sdk/wiki/Getting-Started#requirements-and-dependencies).**

3. **Cloned GitHub repository**
    
    Alternatively, you can obtain the MoPub SDK source by cloning the git repository:
    
    `git clone git://github.com/mopub/mopub-android-sdk.git`
    
    **For additional integration instructions, please see the [Getting Started guide](https://github.com/mopub/mopub-android-sdk/wiki/Getting-Started#requirements-and-dependencies).**

## New in this Version
Please view the [changelog](https://github.com/mopub/mopub-android-sdk/blob/master/CHANGELOG.md) for a complete list of additions, fixes, and enhancements in the latest release.

- **Features**
  - SDK binary artifacts are uploaded to Maven Central Repository. 
  - Update AndroidX Media2 dependencies from version 1.1.1 to 1.1.3.
  - Improve MoPub SDK modularization with 'mopub-sdk-networking' and 'mopub-sdk-util' libraries.
  - Add ProGuard rules for AudioFocusHandler and MediaPlayer.

- **Bug Fixes**
  - Fix rare audio focus handling issue when application is in the background.

## Requirements

- Android 4.4 (API Version 19) and up (**Updated in 5.11.0**)
- androidx.legacy:legacy-support-v4:1.0.0 (**Updated in 5.9.0**)
- androidx.annotation:annotation:1.1.0 (**Updated in 5.9.0**)
- androidx.appcompat:appcompat:1.1.0 (**Updated in 5.10.0**)
- androidx.recyclerview:recyclerview:1.0.0 (**Updated in 5.9.0**)
- androidx.core:core-ktx:1.1.0 (**Added in 5.11.0**)
- com.google.code.gson:gson:2.8.6 (**Added in 5.11.0**)
- androidx.media2:media2-session:1.1.3 (**Updated in 5.17.0**)
- androidx.media2:media2-widget:1.1.3 (**Updated in 5.17.0**)
- androidx.media2:media2-player:1.1.3 (**Updated in 5.17.0**)
- com.mopub:omsdk-android:1.3.16 (**Updated in 5.15.0**)
- MoPub Volley Library (mopub-volley-2.1.0.jar - available on Maven Central) (**Updated in 5.17.0**)
- **Recommended** Google Play Services (com.google.android.gms:play-services-ads-identifier:17.0.0 and com.google.android.gms:play-services-base:17.5.0) (**Updated in 5.15.0**)
- If you are integrating with v5.6.0 or later of the MoPub SDK, specify the `sourceCompatibility` and `targetCompatibility` as below to prevent compilation errors with ExoPlayer 2.9.5 and later:

    ```groovy
    android {
        compileOptions {
            sourceCompatibility JavaVersion.VERSION_1_7
            targetCompatibility JavaVersion.VERSION_1_8
        }
    }
    ```

## Upgrading to SDK 5.13.0
AVID and Moat have been removed as MoPub works on a future viewability solution. No action is necessary, though the maven repository for Moat is no longer required.
All supported network adapters have been updated and are not backwards compatible. Please update to the latest network adapters when upgrading to SDK 5.13.0.

## Upgrading to SDK 5.0

Please see the [Getting Started Guide](https://developers.mopub.com/docs/android/getting-started/) for instructions on upgrading from SDK 4.X to SDK 5.0.

For GDPR-specific upgrading instructions, also see the [GDPR Integration Guide](https://developers.mopub.com/docs/publisher/gdpr-guide).

## Working with Android 6.0 Runtime Permissions
If your app's target SDK is 23 or higher _**and**_ the user's device is running Android 6.0 or higher, you are responsible for supporting [runtime permissions](http://developer.android.com/training/permissions/requesting.html), one of the [changes](http://developer.android.com/about/versions/marshmallow/android-6.0-changes.html) introduced in Android 6.0 (API level 23). In addition to listing any dangerous permissions your app needs in the manifest, your app also has to explicitly request the dangerous permission(s) during runtime by calling method `requestPermissions()` in the [`ActivityCompat`](http://developer.android.com/reference/android/support/v4/app/ActivityCompat.html) class.

### Specifically for the MoPub SDK:
- Dangerous permission [`ACCESS_COARSE_LOCATION`](http://developer.android.com/reference/android/Manifest.permission.html#ACCESS_COARSE_LOCATION) is needed to pass network location data to MoPub.
- Dangerous permission [`ACCESS_FINE_LOCATION`](http://developer.android.com/reference/android/Manifest.permission.html#ACCESS_FINE_LOCATION) is needed to pass GPS location data to MoPub.
    - Granting `ACCESS_FINE_LOCATION` also allows network location data to be passed to MoPub without the need to also grant `ACCESS_COARSE_LOCATION`.
- _**Note:** SDK Version 5.9.0 removes the requirement for the dangerous permission [`WRITE_EXTERNAL_STORAGE`](http://developer.android.com/reference/android/Manifest.permission.html#WRITE_EXTERNAL_STORAGE)._
- _**Note:** The user can deny granting any dangerous permissions during runtime, so please make sure your app can handle this properly._
- _**Note:** The user can revoke any permissions granted previously by going to your app's Settings screen, so please make sure your app can handle this properly._

### Additional resources:
- [Android 6.0 Changes](http://developer.android.com/about/versions/marshmallow/android-6.0-changes.html)
- [Requesting Permissions at Run Time](http://developer.android.com/training/permissions/requesting.html)
- [Permissions Best Practices](http://developer.android.com/training/permissions/best-practices.html)
- [Normal vs Dangerous Permissions](http://developer.android.com/guide/topics/security/permissions.html#normal-dangerous)
- [Permission Groups](http://developer.android.com/guide/topics/security/permissions.html#perm-groups)

## License

To view the full license, visit [http://www.mopub.com/legal/sdk-license-agreement/](http://www.mopub.com/legal/sdk-license-agreement/).

## Open Measurement License

We have partnered with the IAB to provide Viewability measurement via the Open Measurement SDK as of version 5.14.0. To view the full license, visit [https://www.mopub.com/en/omlv1](https://www.mopub.com/en/omlv1)


CHANGELOG.md
## Version 5.17.0 (May 11, 2021)
- **Features**
  - SDK binary artifacts are uploaded to Maven Central Repository. 
  - Update AndroidX Media2 dependencies from version 1.1.1 to 1.1.3.
  - Improve MoPub SDK modularization with 'mopub-sdk-networking' and 'mopub-sdk-util' libraries.
  - Add ProGuard rules for AudioFocusHandler and MediaPlayer.

- **Bug Fixes**
  - Fix rare audio focus handling issue when application is in the background.
    
## Version 5.16.4 (March 31, 2021)
- **Bug Fixes**
  - Fix `InvalidStateException` when finishing VAST videos on some devices.

## Version 5.16.3 (March 24, 2021)
- **Bug Fixes**
  - Handle WebViewClient#onRenderProcessGone for API 26+ devices regarding the VAST icon WebView so WebView crashes do not take the entire process with it.
  - Fix rewarded video duration calculation.
  - Fix rare `InvalidStateException` when finishing a VAST video with a companion image.
  - Fix `UnsupportedOperationException` caused by non-hierarchical URIs.
  - Fix release issues from 5.16.2.

## Version 5.16.2 (March 18, 2021) *Deprecated*
- **Bug Fixes**
  - Handle WebViewClient#onRenderProcessGone for API 26+ devices regarding the VAST icon WebView so WebView crashes do not take the entire process with it.
  - Fix rewarded video duration calculation.
  - Fix rare `InvalidStateException` when finishing a VAST video with a companion image.
  - Fix `UnsupportedOperationException` caused by non-hierarchical URIs.

## Version 5.16.0 (February 16, 2021)
- **Features**
  - Add support for rewarded display ads and rewarded ads with rich media end cards. See [Integration Steps](https://developers.mopub.com/publishers/android/rewarded-ad/).
  - Rename `MoPubRewardedVideos` to `MoPubRewardedAds` and `MoPubRewardedVideoListener` to `MoPubRewardedAdListener`. See the [API Reference](https://developers.mopub.com/publishers/reference/android/MoPub/) for more details.
  - Add 5G cellular support.
  - Update AndroidX Media2 dependencies from version 1.0.1 to 1.1.1.

- **Bug Fixes**
  - Fix crash when adding a theme to native recycler view layout.
  - Fix memory leak when destroy is called on a fullscreen ad.
  - Fix rotation for ads without an orientation specified.
  - Fix rewarded ads being playable after they have expired.
  - Address clickability issues with image and rewarded ads.
  - Address `MoPubRewardedVideos.hasRewardedVideo()` returning the wrong value in certain instances.

## Version 5.15.0 (November 18, 2020)
- **Features**
  - Update countdown animation background color to black for better visibility.
  - Enforce HTTPS for base URLs.
  - Remove native video support.
  - Add support for Snap Audience Network.
  - Update sample app dependency on `com.google.android.gms:play-services-base` from version 17.3.0 to 17.5.0.

## Version 5.14.0 (October 1, 2020)
- **Features**
  - Add beta support for OMSDK version 1.3.4.
  - Certify MoPub Android SDK for Android 11 (API level 30).
  - Add `MoPubErrorCode.TOO_MANY_REQUESTS` to notify of making too many unsuccessful requests in a short period of time.
  - Add Pangle as a supported network.
  - Remove Mintegral as a supported network.

- **Bug Fixes**
  - Banner pause should not restart refresh timer.
  - Address a null pointer exception when some ads expire.
  - Address a destroyed ad causing a null pointer in certain situations.
  - Address having multiple close buttons.
  - Put banner and interstitial callbacks on the main thread. This should address some race conditions when showing an ad immediately after the load finishes.
  - Other minor bugs.

## Version 5.13.1 (July 6, 2020)
- **Bug Fixes**
  - Fix a bug regarding mediated network failovers.

## Version 5.13.0 (June 15, 2020)
- **Features**
  - Remove Moat and IAS measurement SDKs.
  - Consolidate banners, interstitials, and rewarded ads into one container. Third party network adapters for these formats should now extend `BaseAd`.
  - Consolidate the `mopub-sdk-interstitial` and `mopub-sdk-rewarded-video` modules into `mopub-sdk-fullscreen`.
  - Upgrade to use the Androidx Media2 video player for VAST videos.

- **Bug Fixes**
  - Unify the design treatment of fullscreen close buttons. Add a skip button for video when the skip threshold has been met before it has completed.
  - Fix the version name of the sample app on the Play Store.

## Version 5.12.0 (April 6, 2020)
- **Features**
  - Add Mintegral as a supported network.
  - Geographical location is set automatically by the SDK. Public methods to set location are marked deprecated.
  - Add a new method `getAppVersion()` to the `ImpressionData` class.
  - MRAID ads always show native close button.

- **Bug Fixes**
  - Fix multiple UI layout warnings in the sample application.
  - Remove parameter android_perms_ext_storage from an ad request.

## Version 5.11.1 (February 18, 2020)
- **Bug Fixes**
  - Addresses a crash for Android API versions 19-23

## Version 5.11.0 (January 28, 2020)
- **Features**
  - Raise the minimum Android SDK to Android 4.4 (API level 19).
  - Update GDPR logic to allow MoPub to reacquire consent for new vendors.
  - Update our support for OpenRTB Native Ads to version 1.2 and add an optional `sponsored` text field for native ads.
  - Upgrade ExoPlayer to 2.11.0.
  - Add switch to use AndroidX MediaPlayer instead of the default Android MediaPlayer for VAST Video. This adds the dependencies `androidx.media2:media2-session`, `androidx.media2:media2-widget`, and `androidx.media2:media2-player`, all on version 1.0.1.
  - Add dependency on `androidx.core:core-ktx` version 1.1.0.
  - Add dependency on `com.google.code.gson:gson` version 2.8.6.
  - Add the ability to test manual native ads in the sample app.

- **Bug Fixes**
  - Fix an issue with rate limiting for rewarded ads.
  - Fix a compliance issue with Facebook Audience Network native for the sample app.
  - Increase consistency for immersive mode across VAST, MRAID, and HTML ads.

## Version 5.10.0 (October 30, 2019)
- **Features**

mopub-sdk/mopub-sdk-base/src/main/java/com/mopub/common/Constants.java
// Copyright 2018-2020 Twitter, Inc.
// Licensed under the MoPub SDK License Agreement
// http://www.mopub.com/legal/sdk-license-agreement/

package com.mopub.common;

public class Constants {

    private Constants() {}

    static {
        HOST = "ads.mopub.com";
    }

    public static final String HTTP = "http";
    public static final String HTTPS = "https";
    public static final String INTENT_SCHEME = "intent";

    public static final String HOST;

    public static final String AD_HANDLER = "/m/ad";
    public static final String CONVERSION_TRACKING_HANDLER = "/m/open";
    public static final String POSITIONING_HANDLER = "/m/pos";
    public static final String GDPR_SYNC_HANDLER = "/m/gdpr_sync";
    public static final String GDPR_CONSENT_HANDLER = "/m/gdpr_consent_dialog";

    public static final String TAS_AUTHORIZED = "authorized";
    public static final String TAS_DENIED = "denied";

    public static final int TEN_SECONDS_MILLIS = 10 * 1000;
    public static final int THIRTY_SECONDS_MILLIS = 30 * 1000;
    public static final int FIFTEEN_MINUTES_MILLIS = 15 * 60 * 1000;
    public static final int FOUR_HOURS_MILLIS = 4 * 60 * 60 * 1000;

    public static final int AD_EXPIRATION_DELAY = FOUR_HOURS_MILLIS;

    public static final int TEN_MB = 10 * 1024 * 1024;

    public static final int UNUSED_REQUEST_CODE = 255;  // Acceptable range is [0, 255]

    public static final String NATIVE_VIDEO_ID = "native_video_id";
    public static final String NATIVE_VAST_VIDEO_CONFIG = "native_vast_video_config";

    // Android platform
    public static final String ANDROID_PLATFORM = "android";

    // Internal Video Tracking nouns, defined in ad server
    public static final String VIDEO_TRACKING_EVENTS_KEY = "events";
    public static final String VIDEO_TRACKING_URLS_KEY = "urls";
    public static final String VIDEO_TRACKING_URL_MACRO = "%%VIDEO_EVENT%%";

    // VAST JSON Field names
    public static final String VAST_WIDTH = "width";
    public static final String VAST_HEIGHT = "height";
    public static final String VAST_RESOURCE = "resource";
    public static final String VAST_TYPE = "type";
    public static final String VAST_CREATIVE_TYPE = "creative_type";

    public static final String VAST_TRACKER_CONTENT = "content";
    public static final String VAST_TRACKER_MESSAGE_TYPE = "message_type";
    public static final String VAST_TRACKER_REPEATABLE = "is_repeatable";
    public static final String VAST_TRACKER_TRACKING_MS = "tracking_ms";
    public static final String VAST_TRACKER_TRACKING_FRACTION = "tracking_fraction";
    public static final String VAST_TRACKER_PLAYTIME_MS = "playtime_ms";
    public static final String VAST_TRACKER_PERCENT_VIEWABLE = "percent_viewable";

    public static final String VAST_TRACKERS_IMPRESSION = "impression_trackers";
    public static final String VAST_TRACKERS_FRACTIONAL = "fractional_trackers";
    public static final String VAST_TRACKERS_ABSOLUTE = "absolute_trackers";
    public static final String VAST_TRACKERS_PAUSE = "pause_trackers";
    public static final String VAST_TRACKERS_RESUME = "resume_trackers";
    public static final String VAST_TRACKERS_COMPLETE = "complete_trackers";
    public static final String VAST_TRACKERS_CLOSE = "close_trackers";
    public static final String VAST_TRACKERS_SKIP = "skip_trackers";
    public static final String VAST_TRACKERS_CLICK = "click_trackers";
    public static final String VAST_TRACKERS_ERROR = "error_trackers";

    public static final String VAST_URL_CLICKTHROUGH = "clickthrough_url";
    public static final String VAST_URL_NETWORK_MEDIA_FILE = "network_media_file_url";
    public static final String VAST_URL_DISK_MEDIA_FILE = "disk_media_file_url";
    public static final String VAST_SKIP_OFFSET = "skip_offset";
    public static final String VAST_SKIP_OFFSET_MS = "skip_offset_ms";
    public static final String VAST_DURATION_MS = "duration_ms";
    public static final String VAST_COMPANION_AD_LANDSCAPE = "landscape_companion_ad";
    public static final String VAST_COMPANION_AD_PORTRAIT = "portrait_companion_ad";
    public static final String VAST_ICON_CONFIG = "icon_config";
    public static final String VAST_IS_SKIPPABLE = "is_skippable";
    public static final String VAST_IS_REWARDED = "is_rewarded";
    public static final String VAST_ENABLE_CLICK_EXP = "enable_click_exp";

    public static final String VAST_CUSTOM_TEXT_CTA = "custom_cta_text";
    public static final String VAST_CUSTOM_TEXT_SKIP = "custom_skip_text";
    public static final String VAST_CUSTOM_CLOSE_ICON_URL = "custom_close_icon_url";
    public static final String VAST_VIDEO_VIEWABILITY_TRACKER = "video_viewability_tracker";

    public static final String VIEWABILITY_VERIFICATION_RESOURCES = "viewability-verification-resources";

    public static final String VAST_DSP_CREATIVE_ID = "dsp_creative_id";
    public static final String VAST_PRIVACY_ICON_IMAGE_URL = "privacy_icon_image_url";
    public static final String VAST_PRIVACY_ICON_CLICK_URL = "privacy_icon_click_url";

}
